### DISCUSSION TITLE
From traced array to numpy array

### DISCUSSION THREAD

**Discussion by mattiadutto** [jax-ml/jax#12969] [Q&A] [answered]
I'm using Jax for a Federated Learning project, I want to check a few values inside different traced arrays and compere with values inside numpy arrays. So I have the following questions:

- There is any way to move from traced array to a numpy array?
- Or a way to get access to the real value inside a cell of the traced array? 

The best that I found is `array.at[index].get()` but unfortunately it return a traced array.

Thanks. 
Mattia

**Comment by jakevdp** [comment, marked as answer]
*Edit: there is now a FAQ entry on this topic: https://jax.readthedocs.io/en/latest/faq.html#how-can-i-convert-a-jax-tracer-to-a-numpy-array*

---

Thanks for the question! You can print traced values at runtime using [`jax.debug.print`](https://jax.readthedocs.io/en/latest/debugging/print_breakpoint.html). Note however that the syntax of `debug.print` is different than that of normal print statements: the first argument must be a format string. It might look something like this:
```python
import jax
import jax.numpy as jnp

@jax.jit
def f(x):
  print("traced:", x[0])
  jax.debug.print("debug: {}", x[0])

f(jnp.arange(10))
```
```
traced: Traced<ShapedArray(int32[])>with<DynamicJaxprTrace(level=0/1)>
debug: 0
```

**Reply by mattiadutto** [reply]
Thanks for the quick answer!
So i can use `jax.debug.print()` only inside a compiled **jit** function. 

For the part how to move to a Numpy array do you have any suggestions?

**Reply by jakevdp** [reply]
Sorry - I think I missed that part of the question.

You cannot convert a tracer into a numpy array. But if you want to execute numpy code in the midst of a JAX execution, you can do so via a callback to the host using [`jax.pure_callback`](https://jax.readthedocs.io/en/latest/_autosummary/jax.pure_callback.html#jax.pure_callback). It's fairly efficient on CPU, where the host-to-device memory transfers can usually be done in a zero-copy manner, but be aware that on GPU the device transfer overhead and computational lock can cause severe performance hits.

There's not great documentation on this yet, but there's a short demo in this answer: https://github.com/google/jax/discussions/12180#discussioncomment-3517480

**Reply by mattiadutto** [reply]
Don't worry, thanks!
